### PR TITLE
Analyze: add cert check for gateway credential

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
part of https://github.com/istio/istio/issues/43891

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [x] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/config/analysis/analyzers/gateway/secret.go [hunk 1]
```diff
@@ -79,20 +84,64 @@ func (a *SecretAnalyzer) Analyze(ctx analysis.Context) {
...
 		return true
 	})
 }
 
+func isValidSecret(secret *resource.Instance) bool {
+	s, ok := secret.Message.(*corev1.Secret)
+	if !ok {
+		return false
+	}
+	certData, ok := s.Data[ca.TLSSecretCACertFile]
```
Comment (howardjohn) [inline pkg/config/analysis/analyzers/gateway/secret.go:117]: This is not accurate, we have more types of valid secrets.

See extractKeyCertAndStaple

Comment (hanxiaop) [reply pkg/config/analysis/analyzers/gateway/secret.go]: Thanks, revised. I also noticed that there was already a verification logic, and I improved and reused it to ensure consistency.

Comment (hanxiaop) [reply pkg/config/analysis/analyzers/gateway/secret.go]: @howardjohn PTAL, thanks

File: pilot/pkg/xds/sds.go [hunk 2]
```diff
@@ -214,13 +214,38 @@ func (s *SecretGen) generate(sr SecretResource, configClusterSecrets, proxyClust
...
+
+		// add the certificate to the trusted roots pool
+		trustedRoots.AddCert(cert)
+	}
+
+	// verify the certificate chain
+	opts := x509.VerifyOptions{
+		Roots: trustedRoots,
+	}
+	if _, err = certs[0].Verify(opts); err != nil {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/xds/sds.go:245]: also do we really need this?

**Review by howardjohn — COMMENTED** [review]
cc @hzxuzhonghu

File: pilot/pkg/xds/sds.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pilot/pkg/xds/sds.go]: Looks like it's not needed. Right now it only checks whether the cert is in the valid time range.

**Comment by hanxiaop** [comment]
@hzxuzhonghu PTAL, thanks

**Comment by hanxiaop** [comment]
Kindly ping @hzxuzhonghu

File: pkg/config/analysis/analyzers/analyzers_test.go [hunk 3]
```diff
@@ -170,6 +170,17 @@ var testGrid = []testCase{
...
 	},
+	{
+		name:       "gatewaySecretValidations",
+		inputFiles: []string{"testdata/gateway-secrets-validation.yaml"},
+		analyzer:   &gateway.SecretAnalyzer{},
+		expected: []message{
+			{msg.InvalidGatewayCredential, "Gateway defaultgateway-invalid-keys"},
+			{msg.InvalidGatewayCredential, "Gateway defaultgateway-missing-keys"},
+			{msg.InvalidGatewayCredential, "Gateway defaultgateway-invalid-cert"},
+			{msg.InvalidGatewayCredential, "Gateway defaultgateway-expired"},
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/config/analysis/analyzers/analyzers_test.go:181]: The valid period is when the current time falls within the `NotBefore` and `NotAfter` time range of the cert. If the current time is before `NotBefore` or after `NotAfter`, the cert is considered not yet valid or expired.

File: releasenotes/notes/43921-2.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+apiVersion: release-notes/v2
+kind: feature
+area: security
+releaseNotes:
+  - |
+    **Added** additional certificate validations for the `VERIFY_SDS_CERTIFICATE` feature when fetching certificates from the SDS server.
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply releasenotes/notes/43921-2.yaml]: Removed, thanks

File: pkg/config/analysis/analyzers/analyzers_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/config/analysis/analyzers/analyzers_test.go:181]: I mean the cert you introduced

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**